AMR Context Detection
To extract AMR genes and their neighborhood

------------------------------------------------------------------

If this file does not answer your questions, please
feel free to e-mail any questions to Somayeh at:

    somayeh.kafaie@dal.ca

------------------------------------------------------------------


DIRECTORIES:
    -full_pipeline: 			The main file to extract the neighborhood of AMR genes 						from an assembly graph,
					annotate it and compare with the reference genome(s) 
					  
    -extract_neighborhood:		Functions to extract the neighborhood of an AMR gene from 						an assembly graph
    
    -find_seq_in_contigs:		To check if the assembler has generated a contig for an 						extracted neighborhood sequence
    
    -annotation_visualization:	To visualize sequences annotations and make their 						comparison easier
    
    -params:				List of all parameters to be set

TOOLS TO INSTALL:
Please make sure to install this tools before running the code:
	- ART		(available through https://www.niehs.nih.gov/research/resources/software/biostatistics/art/index.cfm) 
	- MetaSPAdes	(available through https://cab.spbu.ru/software/spades/)
	- Bandage	(available through https://rrwick.github.io/Bandage/)
	- Blastn	(available through https://www.ncbi.nlm.nih.gov/books/NBK279671/)
	- Prokka*	(available through https://github.com/tseemann/prokka)
	- RGI		(available through https://github.com/arpcard/rgi)
*In current implementation, Prokka has been run via docker. If you would like to install it from a different channel, you need to update PROKKA_COMMAND_PREFIX variable in params.py with an appropriate value which is probably an empty string.
  
PACKAGES TO INSTALL:
    -numpy         		(available through anaconda or pip)
    -dna_features_viewer	(available through pip or https://github.com/Edinburgh-Genome-Foundry/DnaFeaturesViewer)
    -matplotlib    		(available through anaconda or pip)
    -gfapy			(available through bioconda or pip or https://github.com/ggonnella/gfapy)
    -Bio		       (available through pip)
    -Pillow		       (available through pip)

    *Easiest way to install all required packages is through requirements.txt
     file with the command 'pip install r- requirements.txt'
    **Recommend setting up virtual environment first, so packages and
      versions don't get mixed up with system versions.


==============================================================================

The following arguments can be set via command line or params.py:
optional arguments:
  -h, --help            show this help message and exit
  --task TASK [TASK ...]
                        which task would you like to do? For the entire
                        pipeline choose 0; otherwise either provide a number
                        representing one of the following tasks or two numbers
                        to denote the start and end tasks (and of course all
                        tasks in the middle will be run). Here is the list:
                        metagenome_creation = 1 read_simulation = 2 assembly =
                        3 graph_neighborhood = 4 sequence_neighborhood = 5
                        neighborhood_comparison = 6
  --amr_file AMR_FILE, -A AMR_FILE
                        the path of the file containing the AMR gene sequence
  --ref_genome_files REF_GENOME_FILES [REF_GENOME_FILES ...]
                        the ddress of reference genomes that AMR genome will
                        be inserted in them
  --output_dir OUTPUT_DIR, -O OUTPUT_DIR
                        the output dir to store metagenome file and reads
  --number_of_insertions NUMBER_OF_INSERTIONS
                        the number of genomes generated by inserting AMR in
                        different locations of reference genome
  --insertion_type INSERTION_TYPE
                        Should insertion locations be selected randomly (1) or
                        using some defined values (2)
  --insertion_locations INSERTION_LOCATIONS [INSERTION_LOCATIONS ...]
                        list of predefined insertion locations to insert AMR
                        in reference genome if you chose multiple reference
                        genomes first type the list of insertion locations for
                        the first genome followed by the list for the second
                        and ... please make sure the number of locations for
                        each genome matches the corresponding one specified in
                        --num_insertion argument
  --read_length READ_LENGTH
                        the length of simulated reads can be either 150 or 250
  --spades_thread_num SPADES_THREAD_NUM
                        the number of threads used for MetaSPAdes
  --spades_output_dir SPADES_OUTPUT_DIR
                        the output dir to store MetaSPAdes results
  --graph_distance GRAPH_DISTANCE, -D GRAPH_DISTANCE
                        the maximum distance of neighborhood nodes to be
                        extracted from the AMR gene
  --seq_length SEQ_LENGTH, -L SEQ_LENGTH
                        the length of AMR gene's neighbourhood to be extracted
  --neighborhood_seq_file NEIGHBORHOOD_SEQ_FILE
                        the address of the file containing all extracted
                        neighborhood sequences in assembly graph
  --gfa_file GFA_FILE   the address of the file for assembly graph
  --contig_file CONTIG_FILE
                        the address of the file containing contigs after
                        assembly
  --genome_amr_files GENOME_AMR_FILES [GENOME_AMR_FILES ...]
                        the address of the files containing genome after AMR
                        insertion
  --read1 READ1         the address of the files containing the first file in
                        paired-end reads
  --read2 READ2         the address of the files containing the second file in
                        paired-end reads
  --spades_error_correction SPADES_ERROR_CORRECTION
                        Whether to turn on or off error correction in
                        MetaSPAdes
  --use_RGI USE_RGI     Whether to contribute RGI annotation in Prokka result
  --RGI_include_loose RGI_INCLUDE_LOOSE
                        Whether to include loose cases in RGI result


